api/environments/identities/views.py

```diff
@@ -24,29 +28,112 @@
+            return Key("identifier").eq(search_query.replace('"', ""))
+        else:
+            return Key("identifier").begins_with(search_query)
+
+    @staticmethod
+    def _get_search_kwargs_django(search_query: str) -> dict:
+        filter_kwargs = {}
+        if search_query.startswith('"') and search_query.endswith('"'):
+            filter_kwargs["identifier__exact"] = search_query.replace('"', "")
+        else:
+            filter_kwargs["identifier__icontains"] = search_query
+
+        return filter_kwargs
 
     def get_queryset(self):
         environment = self.get_environment_from_request()
-        queryset = Identity.objects.filter(environment=environment)
 
         search_query = self.request.query_params.get("q")
-        if search_query:
-            if search_query.startswith('"') and search_query.endswith('"'):
-                # Quoted searches should do an exact match just like Google
-                queryset = queryset.filter(
-                    identifier__exact=search_query.replace('"', "")
-                )
-            else:
-                # Otherwise do a fuzzy search
-                queryset = queryset.filter(identifier__icontains=search_query)
+        queryset = Identity.objects.filter(environment=environment)
 
+        if search_query:
+            filter_kwargs = self._get_search_kwargs_django(search_query)
+            queryset = queryset.filter(**filter_kwargs)
         # change the default order by to avoid performance issues with pagination
         # when environments have small number (<page_size) of records
         queryset = queryset.order_by("created_date")
 
         return queryset
 
+    def _get_dynamo_query_kwargs(
+        self,
+        request,
+        environment_api_key: str,
+        previous_last_evaluated_key: str,
+        page_size: int,
+    ) -> dict:
+        search_query = self.request.query_params.get("q")
+
+        filter_expression = Key("environment_api_key").eq(
+            self.kwargs["environment_api_key"]
+        )
+        if search_query:
+            filter_expression = filter_expression & self._get_search_expression_dynamo(
+                search_query
+            )
+
+        dynamo_query_kwargs = {
+            "IndexName": "environment_api_key-identifier-index",
+            "Limit": page_size,
+            "KeyConditionExpression": filter_expression,
+        }
+        if previous_last_evaluated_key:
+            dynamo_query_kwargs.update(
+                ExclusiveStartKey=json.loads(
+                    base64.b64decode(previous_last_evaluated_key)
+                )
+            )
+        return dynamo_query_kwargs
+
+    def list(self, request, *args, **kwargs):
+        environment = self.get_environment_from_request()
+        if not environment.project.enable_dynamo_db:
+            return super().list(request, *args, **kwargs)
+
+        paginator = self.pagination_class()
```

Comment:
We should probably encapsulate the code from here downwards in a well named private method.

```python
def list(self, request, *args, **kwargs):
        environment = self.get_environment_from_request()
        if not environment.project.enable_dynamo_db:
            return super().list(request, *args, **kwargs)

        return self._get_list_response_from_dynamo_records(request, *args, **kwargs)  # or something like this
```